AlgebraInLean/World01/Sheet6.lean

```diff
@@ -0,0 +1,328 @@
+          have hfl : (false, b.2.add c.2).1 = false := rfl
+          rw [h_no_ref_f hfl]
+          simp
+          have h : ∀ a b c : Dn n, (a.2.add b.2).add c.2 = a.2.add (b.2.add c.2)
+          exact fun a b c => add_assoc a.2 b.2 c.2
+          specialize h a b c
+          rw [h]
+    exact fun a => h_exact a
+
+  /- The identity of Dn is the element that does not rotate or reflect the polygon-/
+  id := {1 := false, 2:= 0}
+
+  /- Similarly to the proof of op_assoc, op_id splits into two cases, the applies
+  the theorems shown earlier on each case to simplify before lastly showing that
+  the goal is a theorem already known about the natural numbers. -/
+  op_id := by
+    have h : ∀ a : Dn n, (Dn_op n) a (false, 0) = a
+    intro a
+    have htf := Bool.eq_false_or_eq_true a.1
+    cases htf with
+    | inl ht =>
+      rw [h_op_t ht]
+      have hf2 : (false, 0).1 = false := rfl
+      rw [h_ref_f hf2]
+      simp
+      have hzero : a.2.sub 0 = a.2 := sub_zero a.2
+      rw [hzero, ht.symm]
+      rfl
+    | inr hf =>
+      rw [h_op_f hf]
+      have hf2 : (false, 0).1 = false := rfl
+      rw [h_no_ref_f hf2]
+      simp
+      have hzero : a.2.add 0 = a.2 := add_zero a.2
+      rw [hzero, hf.symm]
+      rfl
+    exact fun a => h a
+
+  /- id_op is shown in the same way as op_id. -/
+  id_op := by
+    have h : ∀ a : Dn n, (Dn_op n) (false, 0) a = a
+    intro a
+    have hf : (false, 0).1 = false := rfl
+    rw [h_op_f hf]
+    have htf := Bool.eq_false_or_eq_true a.1
+    cases htf with
+    | inl ht =>
+      rw [h_no_ref_t ht]
+      simp
+      have hzero1 : Fin.add 0 a.2 = 0 + a.2 := rfl
+      have hzero2 : 0 + a.2 = a.2 := zero_add a.2
+      rw [hzero1, hzero2, ht.symm]
+      rfl
+    | inr hf =>
+      rw [h_no_ref_f hf]
+      simp
+      have hzero1 : Fin.add 0 a.2 = 0 + a.2 := rfl
+      have hzero2 : 0 + a.2 = a.2 := zero_add a.2
+      rw [hzero1, hzero2, hf.symm]
+      rfl
+    exact fun a => h a
+
+  /- The inverse function is the same as defined earlier-/
+  inv := Dn_inv n
+
+  /- inv_op is similar to the other proofs: splitting into cases, simplifing, then
+  figuring out what property of the natural numbers it is equivalent to-/
+  inv_op := by
+    have h : ∀ a : Dn n, (Dn_op n) (Dn_inv n a) a = (false, 0)
+    intro a
+    have htf := Bool.eq_false_or_eq_true a.1
+    cases htf with
+    | inl ht =>
+      rw [h_inv_t ht]
+      have ht2 : (true, a.2).1 = true := rfl
+      rw [h_op_t ht2, h_ref_t ht]
+      simp
+      have hinv1 : a.2.sub a.2 = a.2 - a.2 := rfl
+      have hinv2 : a.2 - a.2 = 0 := sub_eq_zero_of_eq rfl
+      rw [hinv1, hinv2]
+    | inr hf =>
+      rw [h_inv_f hf]
+      have hf2 : (false, -a.2).1 = false := rfl
+      rw [h_op_f hf2, h_no_ref_f hf]
+      simp
+      have hinv1 : (-a.2).add a.2 = (-a.2) + a.2 := rfl
+      have hinv2 : (-a.2) + a.2 = 0
+      exact neg_add_self a.2
+      rw [hinv1, hinv2]
+    exact fun a => h a
+
```

Comment:
Is this meant to be reading material, or will there be a few exercises here and there?